Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.cpp

```diff
@@ -108,85 +132,98 @@ UnacceleratedBuffer::~UnacceleratedBuffer()
     }
 }
 
-void UnacceleratedBuffer::beginPainting()
+#if USE(SKIA)
+Ref<Buffer> AcceleratedBuffer::create(Ref<BitmapTexture>&& texture, Flags flags)
 {
-    Locker locker { m_painting.lock };
-    ASSERT(m_painting.state == PaintingState::Complete);
-    m_painting.state = PaintingState::InProgress;
+    return adoptRef(*new AcceleratedBuffer(WTFMove(texture), flags));
 }
 
-void UnacceleratedBuffer::completePainting()
+AcceleratedBuffer::AcceleratedBuffer(Ref<BitmapTexture>&& texture, Flags flags)
+    : Buffer(flags)
+    , m_texture(WTFMove(texture))
 {
-    Locker locker { m_painting.lock };
-    ASSERT(m_painting.state == PaintingState::InProgress);
-    m_painting.state = PaintingState::Complete;
-    m_painting.condition.notifyOne();
 }
 
-void UnacceleratedBuffer::waitUntilPaintingComplete()
+AcceleratedBuffer::~AcceleratedBuffer()
 {
-    Locker locker { m_painting.lock };
-    m_painting.condition.wait(m_painting.lock, [this] {
-        return m_painting.state == PaintingState::Complete;
+    m_fence = nullptr;
+
+    ensureOnMainThread([this, texture = WTFMove(m_texture)]() mutable {
+        PlatformDisplay::sharedDisplay().skiaGLContext()->makeContextCurrent();
+        texture = nullptr;
```

Comment:
Good catch -- this is in fact completly unnecessary, because the texture that's passed to AcceleratedBuffer, is always kept alive in the BitmapTexturePool owner by SkiaThreadedPaintingPool. So it should always have refCount==2 here. I think we should remove the main-thread destruction and rely on the BitmapTexturePool purging or reusing the texture later.

I will add an assertion to verify refCount == 2 here, ok?

Reply:
As expected, refCount is 2 here, no need to go to main thread.